### ISSUE TITLE
Support `const` statements in `Runner::guess_url()`

### ISSUE THREAD

**Issue by rnaby** [wp-cli/wp-cli#1596]
When I give any comman, wp-cli works fine but gives me this error notice... how can I solve this? I use Ubuntu 14.04.

<code>PHP Notice:  Undefined index: HTTP_HOST in phar:///usr/local/bin/wp/php/wp-cli.php(23) : eval()'d code on line 28</code>

**Comment by DaveHewy** [comment]
Do you have some code in the wp-config file which looks at the HTTP_HOST in $_SERVER ?

**Comment by dnaber-de** [comment]
I've got a similar problem here: fresh WP 4.1 multisite install. (Ubuntu 12.04) 

```
$ wp core check-update
PHP Notice:  Undefined index: HTTP_HOST in /var/www//***/wp-includes/ms-settings.php on line 30
Notice: Undefined index: HTTP_HOST in /var/www/***/wp-includes/ms-settings.php on line 30
```

```
$ wp cli info
PHP binary:	
PHP version:	5.3.10-1ubuntu3.15
php.ini used:	/etc/php5/cli/php.ini
WP-CLI root dir:	phar://wp-cli.phar
WP-CLI global config:	
WP-CLI project config:	
WP-CLI version:	0.17.1
```

The `wp-config.php`:
```
<?php
/**
 * MySQL credentials
 */
const DB_NAME     = '******';
const DB_USER     = '******';
const DB_PASSWORD = '******';
const DB_HOST     = 'localhost';
const DB_CHARSET  = 'utf8';
const DB_COLLATE  = '';

$GLOBALS[ 'table_prefix' ] = '******_';

/**
 * multisite
 */
const WP_ALLOW_MULTISITE    = TRUE;
const MULTISITE             = TRUE;
const DOMAIN_CURRENT_SITE  = '******.dev';
const PATH_CURRENT_SITE    = '/';
const SITE_ID_CURRENT_SITE = 1;
const SUBDOMAIN_INSTALL    = TRUE;
#const COOKIE_DOMAIN        = '';


/**
 * debugging
 */
const WP_DEBUG         = TRUE;
const WP_DEBUG_DISPLAY = TRUE;
const WP_DEBUG_LOG     = FALSE;
const SCRIPT_DEBUG     = TRUE;

/**
 * security keys
 *
 * @link https://api.wordpress.org/secret-key/1.1/salt/
 */
define('AUTH_KEY',         '******');
define('SECURE_AUTH_KEY',  '******');
define('LOGGED_IN_KEY',    '******');
define('NONCE_KEY',        '******');
define('AUTH_SALT',        '******');
define('SECURE_AUTH_SALT', '******');
define('LOGGED_IN_SALT',   '******');
define('NONCE_SALT',       '******');

if ( ! defined( 'ABSPATH' ) )
	define( 'ABSPATH', __DIR__ . '/' );

require_once ABSPATH . 'wp-settings.php';
```

**Comment by rnaby** [comment]
Yes. I've modified the folder structure to make version management easy. @DaveHewy . Is this causing the problem ?

**Comment by danielbachhuber** [comment]
@rnaby @dnaber-de Were you able to track this down?

**Comment by dnaber-de** [comment]
Able maybe but out of free time. At least not before the end of the week.

**Comment by szepeviktor** [comment]
Maybe [here](https://github.com/WordPress/WordPress/blob/master/wp-includes/ms-settings.php#L30) ?

**Comment by szepeviktor** [comment]
```
On a multisite installation, you need to pass a --url parameter, so that WP-CLI knows which site it’s supposed to be operating on...
```

**Comment by szepeviktor** [comment]
You are using `const DOMAIN_CURRENT_SITE` not `define()`.

**Comment by dnaber-de** [comment]
Well, I won't believe it but @szepeviktor pointed it out. The `Runner::guess_url()` parses the `wp-config.php` and [looks for `define` statements](https://github.com/wp-cli/wp-cli/blob/master/php/WP_CLI/Runner.php#L214):

```
$re_define = "#.*define\s*\(\s*(['|\"]{1})(.+)(['|\"]{1})\s*,\s*(['|\"]{1})(.+)(['|\"]{1})\s*\)\s*;#iU";
```

Do you want it to be fixed? It should also look up for `const` statements as this is also a valid method to define constants.

**Comment by szepeviktor** [comment]
Please consider replacing your `const`-s with `define`-s.
http://codex.wordpress.org/Editing_wp-config.php

**Comment by szepeviktor** [comment]
`wp-config.php` seems to me like (I don't remember the english word for programming by functions)
http://php.net/manual/en/language.oop5.constants.php

**Comment by dnaber-de** [comment]
Using `const Foo = 'Bar';` is semantically identical to `define( 'Foo', 'Bar' );` yet better readable and maintainable. However, it is supported since PHP 5.3.0 and as WP-CLI requires at least PHP 5.3.2 one could expect that it is aware for the `const` keyword. 

If you want to have this issue fixed, I would work on a fixture for that and provide a pull request.

**Comment by szepeviktor** [comment]
I don't like many parts of PHP's syntax neither.

**Comment by sebastiaandegeus** [comment]
`const Foo = 'Bar'` and `define( 'Foo', 'Bar' )` is [not](http://php.net/manual/en/language.constants.syntax.php#example-128) the same. 

You have to be careful and think when to use either `define` or `const`

**Comment by szepeviktor** [comment]
@sebastiaandegeus Thanks.

**Comment by dnaber-de** [comment]
Well, I see [no reason](http://stackoverflow.com/a/3193704/2169046) why the these WP constants need to be defined at run-time. Beside that, it's a small performance benefit.

I'm not sure what we are talking about here. The underlying problem is that the parsing of `wp-content.php` is not really solid. So the regular expression also detects constants which are defined inside comments. It wouldn't take much effort to make it at least aware of the `const` keyword. 

However, I offered my help, and I still do. Meanwhile my `wp-config.php` looks like this:

```
const WP_ALLOW_MULTISITE   = TRUE;
const MULTISITE            = TRUE;
const DOMAIN_CURRENT_SITE  = '*********';
const PATH_CURRENT_SITE    = '/';
const SITE_ID_CURRENT_SITE = 1;
const BLOG_ID_CURRENT_SITE = 1;
const SUBDOMAIN_INSTALL    = TRUE;
const COOKIE_DOMAIN        = '';
/**
 * fix for WP-CLI
 * @see https://github.com/wp-cli/wp-cli/issues/1596
 *
define( 'DOMAIN_CURRENT_SITE', '*******' );
define( 'PATH_CURRENT_SITE', '/' );
 */
```

**Comment by szepeviktor** [comment]
Very handy fix.

**Comment by danielbachhuber** [comment]
@dnaber-de Here's a bit of the backstory on the problem #1631

**Comment by dnaber-de** [comment]
@danielbachhuber Well the necessity of interpreting the `wp-config.php` without including is, is clear. The most reliable way would be to use a PHP parser instead of regular expressions.

However, I still don't know why it would be a problem to make `Runner::guess_url()` aware of the `const` statement (_additionally_ to the `defnie` function) using the same way like it looks for `define` (reg exp).

**Comment by danielbachhuber** [comment]
> However, I still don't know why it would be a problem to make Runner::guess_url() aware of the const statement

We can do that.

**Comment by danielbachhuber** [comment]
This will be fixed by #2079
